wp_api/src/middleware.rs

```diff
@@ -0,0 +1,364 @@
+                        redirects,
+                        reason,
+                    },
+                    _ => RequestExecutionError::RequestExecutionFailed {
+                        status_code: None,
+                        redirects: None,
+                        reason: RequestExecutionErrorReason::GenericError {
+                            error_message: e.to_string(),
+                        },
+                    },
+                })
+        }
+    }
+}
+
+// MARK: - RetryMiddleware
+
+#[derive(Debug, uniffi::Object)]
+struct RetryMiddleware {
+    max_retries: u32,
+    max_retry_wait_seconds: u64,
+}
+
+#[uniffi::export]
+impl RetryMiddleware {
+    #[uniffi::constructor]
+    fn new(max_retries: u32, max_retry_wait_seconds: u64) -> Self {
+        println!("Creating retry middleware");
+        Self {
+            max_retries,
+            max_retry_wait_seconds,
+        }
+    }
+}
+
+#[uniffi::export]
+#[async_trait::async_trait]
+impl WpApiMiddleware for RetryMiddleware {
+    async fn process(
+        &self,
+        request_executor: Arc<(dyn RequestExecutor + 'static)>,
+        response: WpNetworkResponse,
+        request: Arc<WpNetworkRequest>,
+    ) -> Result<WpNetworkResponse, WpApiError> {
+        let mut retry_count = 0;
+        let mut response = response;
+
+        loop {
+            if retry_count >= self.max_retries {
+                return Err(WpApiError::RequestExecutionFailed {
+                    status_code: Some(response.status_code),
+                    redirects: None,
+                    reason: RequestExecutionErrorReason::MisconfiguredRateLimitError {},
+                });
+            }
+
+            if response.is_rate_limit_exceeded() && self.max_retries != 0 {
+                let retry_after = response.get_retry_after();
+
+                if let Some(mut retry_after) = retry_after {
+                    // If the server sends some super-long value, we don't want to wait that long
+                    if retry_after > self.max_retry_wait_seconds {
+                        retry_after = self.max_retry_wait_seconds;
+                        async_sleep(Duration::from_secs(retry_after)).await;
+                    }
+
+                    // task::sleep(Duration::from_secs(retry_after)).await;
+                    response = request_executor.execute(request.clone()).await?;
+                } else {
+                    return Ok(response); // It's not ok, but we'll let the layer above handle that – we have no idea how long to wait so we shouldn't try
+                }
+                retry_count += 1;
+            }
+        }
+    }
+}
+
+fn async_sleep(duration: std::time::Duration) -> impl Future<Output = ()> {
+    let (tx, rx) = futures::channel::oneshot::channel::<i32>();
+
+    thread::spawn(move || {
+        thread::sleep(duration);
+        let _ = tx.send(0); // ignore error because it is perfectly ok when the receiver is dropped
+    });
```

Comment:
It _feels_ a bit wasteful to create a thread just to let it sleep and do nothing. What do you think about delegating this sleep to the native platforms? 

For example:

```rust
#[uniffi::foreight_trait]
trait RetryMiddlewareSupport {
  async sleep(duration: Duration)
}

struct RetryMiddleware {
  ...
  support: Arc<dyn RetryMiddlewareSupport>
}
```

Native platforms can easily implement `RetryMiddlewareSupport` and pass an instance to `RetryMiddle`'s initializer.

Reply:
This is probably the way to go for sure